internal/controller/streams.go

```diff
@@ -197,105 +165,269 @@ func (sm *streamManager) BootUpStreamAndReconnect(ctx context.Context, cancel co
+			disableCilium <- true
 			continue
-		}
-		result := randomInt.Int64()
-		sleep := 1*time.Second + backoff + time.Duration(result)*time.Millisecond // Add randomness
-		logger.Infow("Failed to establish connection; will retry", "delay", sleep)
-		time.Sleep(sleep)
-		backoff = backoff * 2 // Exponential increase
-		clientID, clientSecret, err := authn.ReadCredentialsK8sSecrets(ctx, envMap.ClusterCreds)
-		if err != nil {
-			logger.Errorw("Could not read K8s credentials", "error", err)
-		}
-		if clientID == "" && clientSecret == "" {
-			OnboardingCredentials, err := authn.GetOnboardingCredentials(ctx, envMap.OnboardingClientId, envMap.OnboardingClientSecret)
-			if err != nil {
-				logger.Errorw("Failed to get onboarding credentials", "error", err)
-				continue
-			}
-			responseData, err := Onboard(ctx, envMap.TlsSkipVerify, envMap.OnboardingEndpoint, OnboardingCredentials, logger)
+		default:
+			err := connectAndStream(logger, sm)
 			if err != nil {
-				logger.Errorw("Failed to register cluster", "error", err)
-				continue
-			}
-			err = authn.WriteK8sSecret(ctx, responseData, envMap.ClusterCreds)
-			time.Sleep(1 * time.Second)
-			if err != nil {
-				logger.Errorw("Failed to write secret to Kubernetes", "error", err)
-			}
-			clientID, clientSecret, err = authn.ReadCredentialsK8sSecrets(ctx, envMap.ClusterCreds)
-			if err != nil {
-				logger.Errorw("Could not read K8s credentials", "error", err)
-				continue
+
+				// If Hubble is not enabled prevent cilium stream from trying to reconnect
+				if errors.Is(err, ErrHubbleNotFound) || errors.Is(err, ErrNoPortsAvailable) {
+					disableCilium <- true
+				}
+
+				logger.Errorw("Failed to establish stream connection; will retry", "error", err)
+				consecutiveFailures++
+
+				if consecutiveFailures >= severeErrorThreshold {
+					close(done)
+					return
+				}
+
+				randomInt, err := rand.Int(rand.Reader, max)
+				if err != nil {
+					logger.Errorw("Could not generate a random int", "error", err)
+					continue
+				}
+
+				jitterPct, _ := randomInt.Float64()
+				jitterPct = jitterPct * maxJitterPct
+				sleep := time.Duration(float64(backoff) * (1. - jitterPct))
+
+				if sleep < initialBackoff {
+					sleep = initialBackoff
+				}
+				if sleep > maxBackoff {
+					sleep = maxBackoff
+				}
+
+				logger.Infow("Sleeping before retrying connection", "backoff", sleep)
+				time.Sleep(sleep)
+				backoff *= 2
+				if backoff > maxBackoff {
+					backoff = maxBackoff
+				}
+			} else {
+				consecutiveFailures = 0
+				backoff = initialBackoff
 			}
 		}
-		conn, err := SetUpOAuthConnection(ctx, logger, envMap.TokenEndpoint, envMap.TlsSkipVerify, clientID, clientSecret)
+	}
+}
+
+// ExponentialStreamConnect will continue to reboot and restart the main operations within the operator if any disconnects or errors occur.
+func ExponentialStreamConnect(ctx context.Context, logger *zap.SugaredLogger, envMap EnvironmentConfig, bufferedGrpcSyncer *BufferedGrpcWriteSyncer) {
```

Comment:
What is `ctx` used for?

Reply:
It is used in NewAuthenticatedCon to get Oauth Token and grpc connection